### DISCUSSION TITLE
Debug outside softconsole via openocd

### DISCUSSION THREAD

**Discussion by Maziar123** [polarfire-soc/polarfire-soc-documentation#138] [Bare metal embedded software]
Hi

I really don't like eclipse :-(

Have any way use softconsole openocd and use VsCode  to debug smartfusion2 & polarfire ??

**Comment by hughbreslin** [comment]
Please follow the steps in the issue template, this issue should be filed as a discussion item.

**Comment by hughbreslin** [comment]
Hi @Maziar123 

This isn't really something we can support here (the organization is for PolarFire SoC). What I can suggest is having a look at the [debugging the HSS from Scratchpad](https://github.com/orgs/polarfire-soc/discussions/100) thread. The section "Step 2 - Use the debugger to download hss-l2scratch.elf to L2 Scratchpad" in the using the command line section (i.e the second instance of this title, I can't link directly to it unfortunately), shows how to call openOCD and connect GDB which you could probably do for SF2 / PF. This is up to you though as its outside the scope here.

**Comment by Maziar123** [comment]
Hi

Yes, I use smartfusion2, but of course after it dive to polarfire !

Thank I read your link, but my problem exactly after load GDB how connect VS Code or clion to it for debug ?

**Comment by Maziar123** [comment]
1. I connect to openocd : ${SC_INSTALL_DIR}/openocd/bin/openocd --command "set DEVICE M2S010" --file board/microsemi-cortex-m3.cfg    

2. connect to gdb ${SC_INSTALL_DIR}/arm-none-eabi-gcc/bin/arm-none-eabi-gdb     

3. in gdb same as link: 
   target remote localhost:3333
   file PATH_TO_ELF
   load PATH_TO_ELF
   layout split
   thread apply all set $pc=_start

But debug start in wrong location and step show something is wrong ..... 
but in the same elf in softconsole all  is normal ...

what forget to do ?

**Reply by AntonKrug** [reply]
![image](https://user-images.githubusercontent.com/30338596/185362795-a4d0fba8-6a57-4819-a87b-a847910ae667.png)

If you want to use the SmartFusion2 outside SoftConsole, then you need to mimic what the launcher is doing, look at what the launchers are doing and figure out what commands you need to pass.

SmartFusion2 launcher doesn't contain the `thread apply all set $pc=_start` and the `layout split` is only for the UI (called TUI) which could be left out when you are expierencing issues.

And my preference instead `file PATH_TO_ELF` is to give the symbols to the GDB before it connects to the target, so step 2 `${SC_INSTALL_DIR}/arm-none-eabi-gcc/bin/arm-none-eabi-gdb PATH_TO_ELF`

Another thing is to consider that thre is custom gdbinit as well `${SC_INSTALL_DIR}/gdbinit/softconsole.gdbinit`

> I really don't like eclipse :-(
> 
> Have any way use softconsole openocd and use VsCode to debug smartfusion2 & polarfire ??

Technically you can use the tools in a way which is not supported, but then you need figure it out how to make it work.

**Reply by Maziar123** [reply]
Hi Anton Krug

Have a clear way to connect & use bundle GDB or system GDB to debug smartfusion2 ?

**Reply by TommyMurphyTM1234** [reply]
This blog post might be useful for somebody who wants to try to use Visual Studio Code and its Embedded Tools extension with the SoftConsole tools (compiler toolchains, gdb, OpenOCD etc.) to do embedded development.

https://devblogs.microsoft.com/cppblog/vscode-embedded-development/

**Reply by Maziar123** [reply]
Hi, Tommy Murphy, As My last comment, my problem is **not the IDE side**. The bundle & system GDB and bundle openocd can not connect normally to smartfusion2. And it seems software does some magician's way (such as using gdbinit or some other !) **without documenting clearly for the user**
Then users are limited to using softconsole only !! I think it is unusual to see other similar chips ….

**Reply by TommyMurphyTM1234** [reply]
I don't know what you mean by "bundle and system gdb"?
But, yes, SoftConsole does use a custom `softconsole.gdbinit` (which is well commented and, if I recall correctly, also mentioned in the SoftConsole documentation), some other gdb commands for setting up a debug connection, and a customised OpenOCD (and fpServer companion utility).
The custom `gdbinit` and gdb commands used are self evident from looking any of the example debug launch configurations provided.
Otherwise, a debug connection involves running OpenOCD (which also runs fpServer to manage the FlashPro connection), and then gdb which connects to OpenOCD's gdb Remote Serial Protocol server to communicate with the debug target over JTAG/FlashPro.
In my opinion, there's nothing that mysterious here.

**Reply by Maziar123** [reply]
_I don't know what you mean by "bundle and system gdb"_
my comment in "on Jul 20" two gdb used both on Linux: one at ${SC_INSTALL_DIR} AND testing with a system(Linux) gdb ver 12.1

_"mentioned in the SoftConsole documentation"_ 
Can give me what location in the document to comment about it?
I have v2021.3 **but can not find any related document and use it** for this job.

As AntonKrug made some comments about gdbinit. Does nobody get a clear sample to debug **SF2** using gdb directly without softconsole IDE ?

**Reply by TommyMurphyTM1234** [reply]
> _I don't know what you mean by "bundle and system gdb"_ my comment in "on Jul 20" two gdb used both on Linux: one at ${SC_INSTALL_DIR} AND testing with a system(Linux) gdb ver 12.1
 
I still don't understand what you mean by "system (Linux) gdb". But, in any case, you need the SoftConsole (or maybe another RISC-V bare metal toolchain build of gdb) to debug here.

> _"mentioned in the SoftConsole documentation"_ Can give me what location in the document to comment about it? I have v2021.3 **but can not find any related document and use it** for this job.

I thought that it might be mentioned here but it's not v

https://mi-v-ecosystem.github.io/SoftConsole-Documentation/SoftConsole-v2022.2-RISC-V/using_softconsole/debugging.html

However, I just recalled that it's actually configured "globally" under _Window > Preferences > C/C++ > Debug > GDB > GDB command file =_ `${eclipse_home}/../gdbinit/SoftConsole.gdbinit`

> As AntonKrug made some comments about gdbinit. Does nobody get a clear sample to debug **SF2** using gdb directly without softconsole IDE ?

Follow the settings in an existing SmartFusion2 debug launch configuration, run OpenOCD with the appropriate options, run gdb with the appropriate commands (including initially sourcing the custom `softconsole.gdbinit`), and connect gdb to OpenOCD to debug the remote target. Have you tried that?
